README.md
# growdev-custom-wc
Demo WooCommerce plugin for twitch stream 


## Goal 1

Default functionality of WooCommerce and WooCommerce Subscriptions 
is to charge Sign-up Fee with first order and not first renewal.

1. Example for Subscription with 5 day free trial, $0 Sign-up fee, $45 per month.

Order | Amount
----|---
Parent Order| $0
First renewal | $45
Second renewal | $45

2. Example for Subscription with 5 day free trial, $8 Sign-up fee, $45 per month.

Order | Amount
----|---
Parent Order| $8
First renewal | $45
Second renewal | $45

3. Custom functoionality desired for Subscription with 5 day free trial, $8 Sign-up fee, $45 per month.

Order | Amount
----|---
Parent Order| $0
First renewal | $45 + $8
Second renewal | $45

Proposed solution:

- Add "Post Trial Fee" field to Product Data
- When Parent Order is placed
	add meta_key   = "_add_fee"
	    meta_value = "true"
- Hook into Renewal Process 
	if (_add_fee)
		add fee from field ($8)
		delete_meta( _add_fee )


changelog.txt
==== GrowDev Custom PHP ====

2020-08-09 0.0.1
* First version



tests/bootstrap.php
<?php
/**
 * PHPUnit bootstrap file
 *
 * @package Growdev_Custom_Wc
 */

$_tests_dir = getenv( 'WP_TESTS_DIR' );

if ( ! $_tests_dir ) {
	$_tests_dir = rtrim( sys_get_temp_dir(), '/\\' ) . '/wordpress-tests-lib';
}

if ( ! file_exists( $_tests_dir . '/includes/functions.php' ) ) {
	echo "Could not find $_tests_dir/includes/functions.php, have you run bin/install-wp-tests.sh ?" . PHP_EOL; // WPCS: XSS ok.
	exit( 1 );
}

// Give access to tests_add_filter() function.
require_once $_tests_dir . '/includes/functions.php';

/**
 * Manually load the plugin being tested.
 */
function _manually_load_plugin() {
	require dirname( dirname( __FILE__ ) ) . '/growdev-custom-wc.php';
}
tests_add_filter( 'muplugins_loaded', '_manually_load_plugin' );

// Start up the WP testing environment.
require $_tests_dir . '/includes/bootstrap.php';


tests/test-sample.php
<?php
/**
 * Class SampleTest
 *
 * @package Growdev_Custom_Wc
 */

/**
 * Sample test case.
 */
class SampleTest extends WP_UnitTestCase {

	/**
	 * A single example test.
	 */
	public function test_sample() {
		// Replace this with some actual testing code.
		$this->assertTrue( true );
	}

	/**
	 * Pointless test to see if this is on.
	 */
	public function test_again() {
		$this->assertTrue( true );
	}
}


growdev-custom-wc.php
<?php
/**
 * Plugin Name:  GrowDev Custom WC
 * Plugin URI:   https://growdevelopment.com/
 * Description:  Custom WooCommerce plugin for demo
 * Version:      0.1
 * Author:       Grow Development
 * Author URI:   https://growdevelopment.com/
*/



add_action( 'init', 'gdcwc_init' );

/**
 * Place to add actions we care about.
 */
function gdcwc_init() {
	// Add post trial fee to products
	add_action( 'woocommerce_subscriptions_product_options_pricing', 'gdcwc_woocommerce_subscriptions_product_options_pricing', 10 );

	// Save post trial fee
	add_action( 'save_post', 'gdcwc_save_subscription_meta', 11 );
}

/**
 * Add Post Trial Fee field to Subscription Pricing
 *
 * @hook 'woocommerce_subscriptions_product_options_pricing'
 */
function gdcwc_woocommerce_subscriptions_product_options_pricing() {
	global $post;

	// Sign-up Fee
	woocommerce_wp_text_input(
		array(
			'id'                => '_subscription_post_trial_fee',
			// Keep wc_input_subscription_intial_price for backward compatibility.
			'class'             => 'wc_input_subscription_intial_price wc_input_subscription_initial_price wc_input_price  short',
			// translators: %s is a currency symbol / code
			'label'             => sprintf( __( 'Post trial fee (%s)', 'woocommerce-subscriptions' ), get_woocommerce_currency_symbol() ),
			'placeholder'       => _x( 'e.g. 9.90', 'example price', 'woocommerce-subscriptions' ),
			'description'       => __( 'Optionally include an amount to be charged after the subscription trial period.', 'woocommerce-subscriptions' ),
			'desc_tip'          => true,
			'type'              => 'text',
			'data_type'         => 'price',
			'custom_attributes' => array(
				'step' => 'any',
				'min'  => '0',
			),
		)
	);
}

/**
 * Save simple subscription schedule box meta
 *
 * @param $post_id
 */
function gdcwc_save_subscription_meta( $post_id ) {

	// TODO: add nonce
	//if ( empty( $_POST['_wcsnonce'] ) || ! wp_verify_nonce( $_POST['_wcsnonce'], 'wcs_subscription_meta' ) ) {
		//return;
	//}
	if ( isset( $_REQUEST['_subscription_post_trial_fee'] ) ) {
		update_post_meta( $post_id, '_subscription_post_trial_fee', $_REQUEST['_subscription_post_trial_fee'] );
	}
	//if ( isset( $_REQUEST['_variable_subscription_schedule_id'] ) ) {
	//	update_post_meta( $post_id, '_variable_subscription_schedule_id', $_REQUEST['_variable_subscription_schedule_id'] );
	//}
}


woocommerce-subscriptions-post-trial-fee.php
<?php
/**
 * Plugin Name:  WooCommerce Subscriptions Post Trial Fee
 * Plugin URI:   https://github.com/growdev/woocommerce-subscriptions-post-trial-fee
 * Description:  Add ability to charge fee after subscription free trial.
 * Version:      0.1
 * Author:       Grow Development
 * Author URI:   https://growdevelopment.com/
 * Textdomain:   'wc-subs-post-trial-fee'
*/


add_action( 'init', 'gdcwc_init' );

/**
 * Place to add actions we care about.
 */
function gdcwc_init() {
	// Add post trial fee to products
	add_action( 'woocommerce_subscriptions_product_options_pricing', 'gdcwc_woocommerce_subscriptions_product_options_pricing', 10 );

	// Save post trial fee
	add_action( 'save_post', 'gdcwc_save_subscription_meta', 11 );

	// Hook into order creation and maybe add post trial fee meta
	add_action( 'woocommerce_checkout_subscription_created', 'gdcwc_woocommerce_checkout_subscription_created', 10, 3 );

	// Hook into renewal order creation and maybe add fee
	add_action( 'wcs_new_order_created', 'gdcwc_maybe_add_fee', 10, 3 );
}

/**
 * Add Post Trial Fee field to Subscription Pricing
 *
 * @hook 'woocommerce_subscriptions_product_options_pricing'
 */
function gdcwc_woocommerce_subscriptions_product_options_pricing() {
	global $post;

	// Sign-up Fee
	woocommerce_wp_text_input(
		array(
			'id'                => '_subscription_post_trial_fee',
			// Keep wc_input_subscription_intial_price for backward compatibility.
			'class'             => 'wc_input_subscription_intial_price wc_input_subscription_initial_price wc_input_price  short',
			// translators: %s is a currency symbol / code
			'label'             => sprintf( __( 'Post trial fee (%s)', 'woocommerce-subscriptions' ), get_woocommerce_currency_symbol() ),
			'placeholder'       => _x( 'e.g. 9.90', 'example price', 'woocommerce-subscriptions' ),
			'description'       => __( 'Optionally include an amount to be charged after the subscription trial period.', 'woocommerce-subscriptions' ),
			'desc_tip'          => true,
			'type'              => 'text',
			'data_type'         => 'price',
			'custom_attributes' => array(
				'step' => 'any',
				'min'  => '0',
			),
		)
	);
}

/**
 * Save simple subscription schedule box meta
 *
 * @param $post_id
 */
function gdcwc_save_subscription_meta( $post_id ) {

	// TODO: add nonce
	//if ( empty( $_POST['_wcsnonce'] ) || ! wp_verify_nonce( $_POST['_wcsnonce'], 'wcs_subscription_meta' ) ) {
	//return;
	//}

	// TODO: Validate this is a float
	if ( isset( $_REQUEST['_subscription_post_trial_fee'] ) ) {
		update_post_meta( $post_id, '_subscription_post_trial_fee', $_REQUEST['_subscription_post_trial_fee'] );
	}
	//if ( isset( $_REQUEST['_variable_subscription_schedule_id'] ) ) {
	//	update_post_meta( $post_id, '_variable_subscription_schedule_id', $_REQUEST['_variable_subscription_schedule_id'] );
	//}
}

/**
 *  Maybe add Post Trial Fee meta to the subscription.
 *
 * @param $subscription    WC_Subscription
 * @param $order           WC_Order
 * @param $recurring_cart  array
 */
function gdcwc_woocommerce_checkout_subscription_created( $subscription, $order, $recurring_cart ) {
	// check order products (line items) for meta
	$items = $order->get_items();

	// TODO: figure out if multiple products have post trial fee.
	foreach ( $items as $item ) {
		$product_id     = $item->get_product_id();
		$post_trial_fee = get_post_meta( $product_id, '_subscription_post_trial_fee', true );
		if ( '' !== $post_trial_fee ) {
			// add to subscription
			update_post_meta( $subscription->get_id(), '_subscription_post_trial_fee', $post_trial_fee );
		}
	}
}

/**
 * Check line items for post trial fee and add if found.
 *
 * @param $new_order        WC_Order
 * @param $subscription     WC_Subscription
 * @param $type             string
 *
 * @return WC_Order
 */
function gdcwc_maybe_add_fee( $new_order, $subscription, $type ) {

	if ( 'renewal_order' !== $type ) {
		return $new_order;
	}

	// Check line items for post trial fee (WC_Order_Item)
	$items = $subscription->get_items();

	foreach ( $items as $item ) {
		$product_id     = $item->get_product_id();
		$post_trial_fee = get_post_meta( $product_id, '_subscription_post_trial_fee', true );
		if ( '' !== $post_trial_fee ) {
			// Create a Fee Object
			$fee           = new StdClass();
			$fee->name     = __( 'Post Trial Fee', 'wc-subs-post-trial-fee' );
			$fee->amount   = $post_trial_fee;
			$fee->taxable  = false;
			$fee->tax      = '';
			$fee->tax_data = '';

			// add fee to order and recalculate
			try {
				$new_order->add_fee( $fee ); // TODO remove deprecated method
				$new_order->save();
				$new_order->calculate_totals();
			} catch ( Exception $e ) {
				// TODO do something
				$new_order->add_order_note( 'Failed to add Post Trial Fee' );
			}
			// remove meta from subscription
			delete_post_meta( $subscription->get_id(), '_subscription_post_trial_fee' );
		}
	}
	return $new_order;
}
